### ISSUE TITLE
A MOOSE input demonstrating several issues

### ISSUE THREAD

**Issue by YaqiWang** [idaholab/moose#9429]
### Description of the enhancement or error report
This simple MOOSE input demonstrates several issues we have seen in Rattlesnake. The input is on 30-line long:
```
[Mesh]
  type = GeneratedMesh
  dim = 3
  nx = 20
  ny = 20
  ny = 20
[]

[Debug]
  show_actions = true
[]

[Executioner]
  type = Steady
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
[]

[Outputs]
  execute_on = 'timestep_end'
  exodus = true
[]

[Testing]
  [./LotsOfDiffusion]
    [./vars]
      number = 1000
    [../]
  [../]
[]
```
Solving the issues is critical to us. But unfortunately they are out of my hand and I spent hours without finding the solution.

### Rationale for the enhancement or information for reproducing the error
What I observed with a debugger by stepping through the codes are:

![image](https://user-images.githubusercontent.com/6607496/27747338-ea7655f8-5d87-11e7-8ef2-047814778313.png)

The issues (ordered by importance) are:
[1] - 319.1MB allocated by PETSc for preconditining, I guess, which is way too high. I know I am using a third party package, HYPRE, but this number still does not make sense.
[2] - Looks like MOOSE allocates 105.1MB memory during the first Jacobian evaluation. Because the memory for Jacobian has been allocated in problem init, this huge chunk of memory is completely unnecessary.
[3] - Problem init allocates too much memory, 221.2MB / (N*8) = 31.3. Considering all vectors added by MOOSE and the Jacobian matrix, this number is reasonable although can be reduced.
[4] - Problem init takes too much time, 30s for a problem with only N=882000 DoFs.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
Rattlesnake performance.

**Comment by YaqiWang** [comment]
Tag @snschune This test basically shows all the problems of your input for the duct paper.

**Comment by permcody** [comment]
Let's have a constructive conversation about this. When I ask a question, it's not to be combative, I genuinely want to understand what's going on just as much as you do.

[1] 319MB for preconditioning ~ 882000/319/8 = 345 Reals per DOF for the preconditioner. Is that a huge number? It's really hard to say. There are two basic classes of data structures, contiguous memory data structures and node-based data structures. We use both a lot but the latter can be pretty expensive. With maps and linked lists you pay an overhead of 2-4x per Real when storing data because of the data structure pointers. With AMG, they may be allocating all of the grids at once so there's several levels of structures. Have you tried with `lu` or other simple preconditioners? We could also experiment with GAMG if we need AMG.
[2] - It's hard to say what the culprit is here. Once we are in the solve, we have the solver memory overhead as well which we know is non-trivial. On the MOOSE side, we also cache a lot of data in maps (remember that overhead as well). We could start pull all of this out but you are always going to see a jump during the solve even if we store no extra data. The good news is that some of that memory is returned between solves.
[3] - Again hard to say where all this memory is at. Problem.init() calls various init() methods down in libMesh so we are absorbing a bunch of allocations and data structures down at that level.
[4] - This is probably worth looking at! It should be easy enough to see what's going on with a profiler. We know that this number is related to local DOFs because it drastically drops when we run with the pre-split meshes. 

Are these numbers all from a serial run?

Also, the other issue is resources. I currently have @rwcarlsen starting work on optimization but he's the only resource on the MOOSE team that can work on this right now. All others are booked (lack of MOOSE-direct funding is difficult).

**Comment by YaqiWang** [comment]
The numbers are from a serial run.

[1] 45 (not 345?) is a huge number. The maximum connectivity of a node in 3D is 27. 45 is much bigger than 27. BTW, this is not in MOOSE. I was told that HYPRE could maintain another copy of the Jacobian due to the different data structure from PETSc. But even considering that, 45 is still a little high. Yes, we can try GAMG.

[2] This memory is allocated by Jacobian evaluation and stays there. Cache cannot use that much memory. There is something not right I think.

[3] I made sure that the memory is used for the Jacobian and bunch of vectors (most of them are not needed for Steady executioner) and the number matches what I expected.

[4] I am also expecting to use the distributed mesh, which I thought should be the default at the beginning. But the number of elements is only 8k, so there could be more we can do.

I am willing to help. For instance, I tried to find what the culprit is for issue 2. But because the offending line must be residing the element loop, I do not have a good mean to detect where the memory is allocated. I deleted few vectors regarding issue 3. Doing that further does not help this test much because the memory is mostly consumed by preconditioning. Even GMRES memory usage (60.7MB, in the lost row of the table) is just 1/10th of the total memory consumption!

**Comment by permcody** [comment]
It's time to spend a little time with massif. That's a valgrind tool for monitoring memory allocations. It'll get you closer to where things are happening. @rwcarlsen has run it, and so have I in the past. It's not an awesome tool by any means but it can help.

Remember that even for a matrix, we are usually storing entries in compressed row format so there's always overhead somewhere because you have pointers to each entry. That might only mean you are off by 2x though, not 10x on your estimates.

**Comment by YaqiWang** [comment]
Is massif available for Mac OS? If the matrix is stored with CSR (compressed sparse row) format, the number of integers as pointers is equal to the number of non-zeros in the matrix. We are using 4 bytes for storing an integer, so 1.5x is more accurate.

**Comment by rwcarlsen** [comment]
In my experience, massif has quesionable reliability on the mac in my experience.  Unless you are using Apple's ``instruments`` for profiling, you should probably do it on linux.

**Comment by YaqiWang** [comment]
Just running the test with `-snes_view` and found the number of nonzeros in the Jacobian is 14884000.

**Comment by rwcarlsen** [comment]
Seems about what I would expect for a 3D with linear shape funcs: ~20 nonzeros per dof.

**Comment by rwcarlsen** [comment]
Just after everything gets set up:

![heap6](https://user-images.githubusercontent.com/1351242/27755198-6acc07fc-5dab-11e7-885f-dfa6f2b94a5c.png)

At the first linear iteration of the first nonlinear iteration:

![heap14](https://user-images.githubusercontent.com/1351242/27755200-6ad1d894-5dab-11e7-90e6-6eb083ebe58f.png)

At the first linear iteration of the second nonlinear iteration:

![heap20](https://user-images.githubusercontent.com/1351242/27755199-6ace3ebe-5dab-11e7-96ac-c8f440cd621f.png)

**Comment by YaqiWang** [comment]
With different solver options, the memory usage is
![image](https://user-images.githubusercontent.com/6607496/27755902-219133aa-5db0-11e7-865f-fa106494e2e6.png)

**Comment by jwpeterson** [comment]
@rwcarlsen these graphs are pretty interesting. Definitely shows that PETSc and Hypre are both holding their own separate copy of the preconditioning matrix... @YaqiWang is also going to add a GAMG line to the table above for comparison...

**Comment by YaqiWang** [comment]
I should mention that in my input, I had a type in the Mesh block. I intended to have `nz = 20` but I copy-paste and duplicated `ny = 20`. Thus the number of elements total is 20*20*1 = 400 but not what I thought 8k. This mistake does not affect the issues but actually makes the input debuggable.

**Comment by YaqiWang** [comment]
@rwcarlsen 's graph is really nice. I need to learn how to generate them on my own. But I can not see the 105MB memory allocated in Jacobian evaluation. I check `opt` executable, the memory consumption is about the same as with `dbg`.

**Comment by YaqiWang** [comment]
I assume the three numbers under PetscMallocAlign are the memory units. 6.73MB is equal to 882099*8 if 1MB=1024*1024, the solution vector size. 173.7MB is about equal to 14884000 * 1.5 * 8, the matrix size. Don't know what is that 3.36MB. PETSc allocates much more vectors than I thought.

**Comment by jwpeterson** [comment]
You may also want to experiment with some other -ksp_types, I think gmres may require more memory than plain old cg, for example.

**Comment by YaqiWang** [comment]
Unfortunately the matrix may not be SPD. Any other suggestions to reduce the memory usage? @rwcarlsen Can you generate a graph with GAMG? Or you can teach me how to do it. Thanks.

**Comment by fdkong** [comment]
> Unfortunately the matrix may not be SPD in general. Any other suggestions to reduce the memory usage? @rwcarlsen Can you generate a graph with GAMG? Or you can teach me how to do it. Thanks.

If the problem is not tough, you can choose a light solver as the subdomain solver such as SOR for GAMG, and this will reduce the memory usage.

**Comment by rwcarlsen** [comment]
@YaqiWang - would you prefer a gamg profile with nz=20 or without to be consistent with the other profiles?

**Comment by YaqiWang** [comment]
@rwcarlsen I do not know what `nz` is. I think to be consistent with other profiles would be fine. @fdkong I am not familiar with GAMG, can you give the exact options? Also is Trilinos solver still available in MOOSE? I'd like to try it out to see if Trilinos will give us a better memory footprint.

**Comment by rwcarlsen** [comment]
@YaqiWang - your typo in the input file for the original profiles that you pointed out hat ``ny=20; ny=20`` instead of ``ny`` and ``nz``.

**Comment by YaqiWang** [comment]
Yes, just change the second `ny = 20` to `nz = 1`. I will update my original post.

**Comment by rwcarlsen** [comment]
Peak memory between 1st nonlinear iteration and next linear iteration:

![heap14](https://user-images.githubusercontent.com/1351242/27885225-77ec9784-6194-11e7-8cb1-3609d6f2087f.png)

At end of 1st linear iteration of 1st nonlinear iteration

![heap19](https://user-images.githubusercontent.com/1351242/27885224-77e40c68-6194-11e7-9254-7ea91542f373.png)

At peak memory between 2nd nonlinear iteration and next linear iteration:

![heap28](https://user-images.githubusercontent.com/1351242/27885226-77ed37f2-6194-11e7-8d6d-d762fc536690.png)

**Comment by YaqiWang** [comment]
I guess Trilinos is no longer a solver option in MOOSE?

**Comment by permcody** [comment]
Not in MOOSE, not for a long time. Trilinos is still supported in libMesh and you can build libMesh with Trilinos support and link it to MOOSE. I'm not sure about how much work it would take to get things running again, probably not a whole lot. The bigger issue is that there's a lot of PETSc specific capabilities that have made their way into MOOSE over the year that you wouldn't be able to leverage with Trilinos.

**Comment by YaqiWang** [comment]
Looks like the recent fix in https://github.com/libMesh/libmesh/pull/1577#issuecomment-365037169 does not reduce the setup time of the small test I attached here. Did I miss something? Tag @roystgnr .

**Comment by roystgnr** [comment]
You see no change at all?  Running that file on 1 core (of an E5-2620, a 5 year old 2GHz CPU) I see
```
 0 Nonlinear |R| = 2.049390e+02
```
after 6 seconds, then job completion after 48 more.

Your table suggests 60 seconds and 180 respectively?

I must be running on a faster processor (or there were other relevant performance fixes in the last 8 months?), since the part of that output which didn't just see a performance fix is still 3.5x faster, but 10x seems decent to me.

**Comment by YaqiWang** [comment]
Nope, I am running with the debug version. Yes, opt version runs much faster. It takes 4s to see
```
0 Nonlinear |R| = 2.049390e+02
```
and total 42s to finish on my desktop. Should I be concerned with the slow setup in debug mode?

**Comment by roystgnr** [comment]
In general, dbg mode is *ridiculously* slow on large problems, and that's not going to change any time soon.  In particular, we turn on the standard glibcxx debugging if it's available, which makes some container operations much slower and makes other container operations *asymptotically* slower.  Our only excuse for keeping it around is that it's successfully automatically caught a host of bugs over the years.

For daily development, I'd recommend the "devel" method of libMesh.  It still gives *most* of the internal debugging support (including all manually written asserts) but arguably doesn't go overboard.  Then when your code appears to be working on test problems you switch to opt to run real problems, or when your code appears to be failing on test problems you switch to dbg to figure out why.

**Comment by YaqiWang** [comment]
Thanks for the info. So it is normal we did not see much gain with the dbg version?

**Comment by roystgnr** [comment]
Hmm... actually, although it is normal not to see much gain in dbg mode from optimizations (the cost of the suboptimal code usually gets dwarfed by the cost of the debugging code), this recent PR wasn't just a performance optimization, it was a performance *fix*, changing the asymptotic performance of some nested loops, and so in this case I'm quite surprised it doesn't affect dbg mode similarly.
